File: src/app/BackEnd/gsb-frais/src/main/java/com/gsb/frais/service/FicheFraisService.java
package com.gsb.frais.service;

import com.gsb.frais.model.FicheFrais;
import com.gsb.frais.repository.FicheFraisRepository;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.web.server.ResponseStatusException;

import java.util.List;

@Service
public class FicheFraisService {

    private final FicheFraisRepository ficheFraisRepository;

    public FicheFraisService(FicheFraisRepository ficheFraisRepository) {
        this.ficheFraisRepository = ficheFraisRepository;
    }

    // Récupérer toutes les fiches de frais
    public List<FicheFrais> getAllFicheFrais() {
        return ficheFraisRepository.findAll();
    }

    // Récupérer une fiche de frais par ID
    public FicheFrais getFicheFraisById(Long id) {
        return ficheFraisRepository.findById(id)
                .orElseThrow(() -> new ResponseStatusException(
                        HttpStatus.NOT_FOUND, "Fiche de frais non trouvée avec l'ID : " + id));
    }

    // Ajouter une nouvelle fiche de frais
    public FicheFrais createFicheFrais(FicheFrais ficheFrais) {
        return ficheFraisRepository.save(ficheFrais);
    }

    // Mettre à jour une fiche de frais
    public FicheFrais updateFicheFrais(Long id, FicheFrais ficheFraisDetails) {
        return ficheFraisRepository.findById(id)
                .map(ficheFrais -> {
                    ficheFrais.setMontantValide(ficheFraisDetails.getMontantValide());
                    ficheFrais.setNbJustificatifs(ficheFraisDetails.getNbJustificatifs());
                    ficheFrais.setDateModif(ficheFraisDetails.getDateModif());
                    return ficheFraisRepository.save(ficheFrais);
                })
                .orElseThrow(() -> new ResponseStatusException(
                        HttpStatus.NOT_FOUND, "Fiche de frais non trouvée avec l'ID : " + id));
    }

    // Supprimer une fiche de frais
    public void deleteFicheFrais(Long id) {
        if (!ficheFraisRepository.existsById(id)) {
            throw new ResponseStatusException(
                    HttpStatus.NOT_FOUND, "Fiche de frais non trouvée avec l'ID : " + id);
        }
        ficheFraisRepository.deleteById(id);
    }
}

File: src/app/Service/FICHEFRAIS/fiche-frais.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { FicheFrais } from '../models/fichefrais.model';  


@Injectable({
  providedIn: 'root'
})
export class FicheFraisService {

  private baseUrl = 'http://localhost:8080/api/fichefrais';  // URL de votre API

  constructor(private http: HttpClient) { }

  // Récupérer toutes les fiches de frais
  getAllFichesFrais(): Observable<FicheFrais[]> {
    return this.http.get<FicheFrais[]>(`${this.baseUrl}`);
  }

  // Ajouter une nouvelle fiche de frais
  createFicheFrais(ficheFrais: FicheFrais): Observable<FicheFrais> {
    return this.http.post<FicheFrais>(`${this.baseUrl}`, ficheFrais);
  }

  // Mettre à jour une fiche de frais
  updateFicheFrais(id: number, ficheFrais: FicheFrais): Observable<FicheFrais> {
    return this.http.put<FicheFrais>(`${this.baseUrl}/${id}`, ficheFrais);
  }

  // Supprimer une fiche de frais
  deleteFicheFrais(id: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/${id}`);
  }
}
